src/docs/memory-management.md

```diff
@@ -0,0 +1,103 @@
+```
+%vtype = alloca %T6output5VTypeV, align 8
+...
+%0 = call ptr @"outlined destroy of output.VType"(ptr %vtype), !dbg !156
+
+// VWT->Destroy
+define linkonce_odr hidden ptr @"outlined destroy of output.VType"(ptr %0) #8 !dbg !143 {
+entry:
+  %.refType = getelementptr inbounds %T6output5VTypeV, ptr %0, i32 0, i32 0, !dbg !144
+  %toDestroy = load ptr, ptr %.refType, align 8, !dbg !144
+  call void @swift_release(ptr %toDestroy) #1, !dbg !144
+  ret ptr %0, !dbg !144
+}
+```
+
+### Initializers
+
+When a reference type or a value type containing a reference property is passed as a parameter to a constructor, its reference count is retained before the call and released after:
+
+```swift
+var refType = RefType()
+var vtype = VType(refType: refType)
+```
+```
+%3 = call ptr @swift_retain(ptr returned %2) #4, !dbg !164
+store ptr %2, ptr %refType, align 8, !dbg !164
+call swiftcc void @"output.VType.init(refType: output.RefType) -> output.VType"(ptr noalias nocapture sret(%T6output5VTypeV) %vtype, ptr %2), !dbg !168
+%toDestroy = load ptr, ptr %refType, align 8, !dbg !170
+call void @swift_release(ptr %toDestroy) #1, !dbg !170
+```
+
+### Parameters
+
+When a reference type or a value type containing a reference property is passed as a parameter, its reference count is retained before the call and released after:
+
+```
+%1 = load ptr, ptr %vtype.refType, align 8, !dbg !154
+%2 = call ptr @swift_retain(ptr returned %1) #4, !dbg !154
+%.refType = getelementptr inbounds %T6output5VTypeV, ptr %0, i32 0, i32 0, !dbg !156
+store ptr %1, ptr %.refType, align 8, !dbg !156
+call swiftcc void @"output.PassThrough(vtype: output.VType) -> ()"(ptr noalias nocapture dereferenceable(8) %0), !dbg !157
+%3 = call ptr @"outlined destroy of output.VType"(ptr %0), !dbg !158
+```
+
+### Return types
+
+When a value type is returned from a function, a new instance is created, and its reference count is retained before the return. The caller is responsible for releasing the reference:
+```
+%2 = load ptr, ptr %.refType, align 8, !dbg !173
+store ptr %2, ptr %vtype.debug, align 8, !dbg !175
+%3 = call ptr @swift_retain(ptr returned %2) #4, !dbg !176
+%.refType1 = getelementptr inbounds %T6output5VTypeV, ptr %0, i32 0, i32 0, !dbg !176
+store ptr %2, ptr %.refType1, align 8, !dbg !176
+```
+
+## Memory handling
+
+To handle native memory in the scenarios above, the projections should could use the value witness table to invoke `InitWithCopy` for copy operations and `Destroy` for finalization. These functions manage reference counts at any level of nesting.
+
+To ensure correct memory handling:
+ - Value types that contain reference properties should be projected as C# classes with destructors that call `Destroy`
+ - When types are marshalled into Swift, `InitWithCopy` should be invoked to create the copy/ `Destroy` should be called on the copy at the end of the block
+ - When types are marshalled from Swift, `Destroy` should be called at the end of the block
```

Comment:
So for C# code like this - it would do something like:
```csharp
ProjectedVType a;

///
// IntPtr r = SwiftPinvoke();
// a = ProjectedVType.NewFromPayload(r);
//     - this internally does a = new ProjectedVType(r)
//           - this in turn effectively does a._payload = InitWithCopy(r);
// Destroy(r);
a = SwiftCallWhichReturnsVType(); 

// Destroy(a._payload)
a.Dispose(); // Or destructor
```

I don't know enough about Swift if this is what will happen in Swift's case as well. C# doesn't have the corresponding "block variable" behavior (object lifetime is not scoped by blocks really). So the equivalent Swift code would be to assign the return value of the function to a field on for example a ref class.

Reply:
Reading this document I get to a conclusion that we do not need `InitWithCopy` when dealing with return values as the reference count will be updated by the callee. Is this right?

Reply:
That was sort of where my question is going to... if Swift can avoid the +/- on return values, maybe we can as well.
Another question related to this is - what are all the ways we can get to an instance of a swift object?
- Return value of a method call
- Create from nothing (driven by us)

The ones which are not covered by this but will become important eventually are:
- Passed in as a parameter of a method implemented in C# but called from Swift
- Accessed as a direct field access? (do we have this? can we actually directly access fields at all anywhere, or do we always go through accessors)

Reply:
>Accessed as a direct field access? (do we have this? can we actually directly access fields at all anywhere, or do we always go through accessors)

We could technically access things directly on things projected as structs - we do not do it now though But then those would be guaranteed to not have ref types (if they had the projection would be a class)

Reply:
Refined the rules, please take a look. 

> Return value of a method call
> Create from nothing (driven by us)

Caller has the ownership and release them at the end of the block.

> Passed in as a parameter of a method implemented in C# but called from Swift

Only if the parameter is returned, a new copy is created/count is retained. Anyway, a caller has the ownership.